src/marqo/tensor_search/tensor_search.py

```diff
@@ -1235,6 +1318,311 @@ def _lexical_search(
+
+        if filter_string is not None:
+            search_query["query"]["nested"]["query"]["knn"][f"{TensorField.chunks}.{vector_field}"][
+                "filter"] = {
+                    "query_string": {"query": f"{contextualised_filter}"}
+                }
+        body += [{"index": index_name}, search_query]
+    return body
+
+def bulk_msearch(config: Config, body: List[Dict]) -> List[Dict]:
+    start_search_http_time = timer()
+    try:
+        response = HttpRequests(config).get(path=F"_msearch", body=utils.dicts_to_jsonl(body))
+        end_search_http_time = timer()
+        total_search_http_time = end_search_http_time - start_search_http_time
+        total_os_process_time = response["took"] * 0.001
+        num_responses = len(response["responses"])
+        logger.debug(f"search (tensor) roundtrip: took {total_search_http_time:.3f}s to send {num_responses} search queries (roundtrip) to Marqo-os.")
+
+        responses = [r['hits']['hits'] for r in response["responses"]]
+
+    except KeyError as e:
+        # KeyError indicates we have received a non-successful result
+        try:
+            if "index.max_result_window" in response["responses"][0]["error"]["root_cause"][0]["reason"]:
+                raise errors.IllegalRequestedDocCount(
+            "Marqo-OS rejected the response due to too many requested results. "
+            "Try reducing the query's limit parameter") from e
+            elif 'parse_exception' in response["responses"][0]["error"]["root_cause"][0]["reason"]:
+                raise errors.InvalidArgError("Syntax error, could not parse filter string") from e
+            raise errors.BackendCommunicationError(f"Error communicating with Marqo-OS backend:\n{response}")
+        except (KeyError, IndexError) as e2:
+            raise e
+
+    logger.debug(f"  search (tensor) Marqo-os processing time: took {total_os_process_time:.3f}s for Marqo-os to execute the search.")
+    return responses
+
+def gather_documents_from_response(resp: List[List[Dict[str, Any]]]) -> Dict[str, Any]:
+    """
+        For the specific responses to a query, gather correct responses. This is used to aggregate documents, for cases
+        where the same document is retrieved for multiple field-queries.
+    """
+    gathered_docs = dict()
+
+    for i, query_res in enumerate(resp):
+        for doc in query_res:
+            doc_chunks = doc["inner_hits"][TensorField.chunks]["hits"]["hits"]
+            if doc["_id"] in gathered_docs:
+                gathered_docs[doc["_id"]]["doc"] = doc
+                gathered_docs[doc["_id"]]["chunks"].extend(doc_chunks)
+            else:
+                gathered_docs[doc["_id"]] = {
+                    "_id": doc["_id"],
+                    "doc": doc,
+                    "chunks": doc_chunks
+                }
+
+        # Filter out docs with no inner hits:
+        for doc_id in list(gathered_docs.keys()):
```

Comment:
nit, we may want to log warnings if this happens (it shouldn't really happen)

Reply:
Again, as above.